fix(tekton): deriving build related meta information from pull refs

hferentschik opened the pull request:
- removing parameters which can be derived from step_create_meta_pipeline
- adding step to merge pull refs in meta pipeline
- making sure the right parameters are passed in step create task in meta pipeline

fixes issue #4551

abayer commented on pkg/cmd/step/create/step_create_meta_pipeline.go:
```diff
@@ -61,14 +57,10 @@ var (
-	SourceURL    string
-	Branch       string
-	PullRefs     string
-	PipelineKind string
```
I'm not sure we can get rid of `PipelineKind`, `Context`, etc as options - but that may just be me thinking about things from the perspective of the `step_create_task` tests.

hferentschik replied:
`Context` is still there, that can indeed not be derived. PipelineKind I derive from the pull ref atm, similar to what the pipeline runner does as well. If need be, this option could be brought back. It seems to me though that right now we only have release or pullrequest. Having two options you can also create inconsitent runs, by for example specifying a pull ref for a pull request, but use kind 'release'. That does not make much sense to me.

abayer replied:
Ah, I misread and just saw the line deleting `Context`, not the new line.

Maybe we should force `pullRequest` as the `Kind` if `PullRefs` is not empty, even if you specify a `Kind`? That would allow `feature` to still work. Or we can just ignore it for now and deal with it later if it comes up. =)

abayer commented on pkg/cmd/step/create/step_create_meta_pipeline.go:
```diff
@@ -291,11 +267,46 @@ func (o *StepCreatePipelineOptions) getClientsAndNamespace() (tektonclient.Inter
+// or 'batch' for Prow batch builds.
+// The method makes the decision purely based on the Prow pull ref. At this stage we don't have the full ProwJov spec
+// available.
+func (o *StepCreatePipelineOptions) determineBranchIdentifier(pullRef prow.PullRefs) string {
```
So all of this locks in to only kicking off builds via prow - that's not going to work. `jx start pipeline` is valid, e.g.

hferentschik replied:
`jx start pipeline` does create a Prow job if prow is enabled and goes into a different branch if not - https://github.com/jenkins-x/jx/blob/65962ff5ef1a6d1c4776daee0163434c9c2cb566/pkg/cmd/start/start_pipeline.go#L156

I think we are fine there.

hferentschik replied:
BTW, this happens in similar form in pipeline runner - https://github.com/jenkins-x/jx/blob/fe66ed51204667e37ed067f65a7ebc67f7c75bbb/pkg/cmd/controller/controller_pipelinerunner.go#L322

abayer replied:
Ok, then we should be fine there.

abayer commented on pkg/cmd/step/create/step_create_meta_pipeline.go:
```diff
@@ -291,11 +267,46 @@ func (o *StepCreatePipelineOptions) getClientsAndNamespace() (tektonclient.Inter
+	if prCount > 0 {
+		kind = jenkinsfile.PipelineKindPullRequest
+	} else {
+		kind = jenkinsfile.PipelineKindRelease
```
How about `feature` pipelines? Not used very often, but it is a supported option.

hferentschik replied:
I meant to bring this up. I have it on my list of open questions. I am not quite sure what `feature` pipelines mean. That said, I don't see how our current code ever allows anything else than `release` or `pull request` atm - https://github.com/jenkins-x/jx/blob/fe66ed51204667e37ed067f65a7ebc67f7c75bbb/pkg/cmd/controller/controller_pipelinerunner.go#L205

Not saying this is correct, I also feel I would like to have at least one other option. I am not fully behind the idea of each merge to master creating a release. I'd like a "master" build occurring when the merge to master happens and then somehow trigger a release build based on some other condition. But right now, it seems we have these two options.

hferentschik replied:
If we want we can bring the explicit pipeline kind option back and pass it through, but from how we currently integrate with Prow it does not really change anything.

abayer replied:
Yeah, the more I think about it, the more I'm leaning towards ignoring `feature` for now. We can revisit it later if we need to.

hferentschik commented:
@abayer, I added some tests and removed the wip status. 

Now it is about addressing any outstanding issues/concerns. 

BTW, something like this should work for you as well to test this locally:
```
./build/jx --verbose step create pipeline --source-url https://github.com/hf-bee/test-go-project.git --job d7cf84b6-9cf8-11e9-8f83-46ab9c3d7824 --default-image hferentschik/jx:dc2b13425 --pull-refs master:79c7e4d,42:89866
```

abayer commented:
/lgtm

Let's do it!

hferentschik commented:
>Let's do it!

+1,  we can adjust later on. But this feels like a cleaner start.
